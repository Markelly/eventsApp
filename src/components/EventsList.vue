<template>
  <div class="events-list">
    <div v-if="isEmptyScreen">No events to show</div>
    <div v-else>
      <SortBy v-if="showAllEvents" @sortEvents="sortEvents" />
      <Card
        v-for="(event, index) in events"
        :key="index"
        :event="event"
        data-card
      />
      <Pagination
        v-if="showPagination"
        :totalPages="totalPages"
        @gotoPage="gotoPage"
      />
    </div>
  </div>
</template>

<script>
import store from "@/store";
import Card from "@/components/Card";
import { SORT_OPTIONS } from "@/utils/constants";
import Pagination from "@/components/Pagination";
import SortBy from "@/components/SortBy";

export default {
  name: "EventsList",
  components: {
    SortBy,
    Pagination,
    Card
  },
  data() {
    return {
      events: [],
      page: 0,
      sort: "",
      totalPages: 0,
      sortOptions: SORT_OPTIONS
    };
  },
  props: {
    showAllEvents: {
      type: Boolean,
      default: false
    }
  },
  async created() {
    this.loadList();
  },
  computed: {
    isEmptyScreen() {
      return !this.events.length;
    },
    showPagination() {
      return this.showAllEvents && this.totalPages > 1;
    }
  },
  methods: {
    async loadList(sort) {
      this.showAllEvents
        ? await store.dispatch("getEvents", { page: this.page, sort })
        : await store.dispatch("getFavorites", { page: this.page, sort });
      this.events = store.state.events;
      this.totalPages = store.state.totalPages;
    },
    gotoPage(page) {
      this.page = page - 1;
      this.loadList();
    },
    sortEvents(sort) {
      this.loadList(sort);
    }
  }
};
</script>

<style scoped lang="scss">
@import "../styles/events-app";
.events-list {
  display: flex;
  flex-direction: column;
  align-items: center;

  & > * {
    width: 100%;
  }
}
</style>
