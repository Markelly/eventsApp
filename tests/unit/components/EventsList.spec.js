import EventsList from "@/components/EventsList";
import { createLocalVue, shallowMount } from "@vue/test-utils";

const localVue = createLocalVue();

describe("<EventsList/>", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(EventsList, {
      localVue,
      propsData: {
        showAllEvents: true
      },
      computed: {
        isEmptyScreen: () => false
      }
    });
  });

  test("should render the component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should load empty list", () => {
    wrapper = shallowMount(EventsList, {
      localVue,
      propsData: {
        showAllEvents: true
      },
      computed: {
        isEmptyScreen: () => true
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  test("should NOT load SortBy components", () => {
    wrapper = shallowMount(EventsList, {
      localVue,
      propsData: {
        showAllEvents: false
      },
      computed: {
        isEmptyScreen: () => false
      }
    });

    expect(wrapper).toMatchSnapshot();
  });

  test("should NOT load SortBy components", () => {
    wrapper = shallowMount(EventsList, {
      localVue,
      data() {
        return { totalPages: 2 };
      },
      propsData: {
        showAllEvents: false
      },
      computed: {
        isEmptyScreen: () => false
      }
    });

    expect(wrapper).toMatchSnapshot();
  });
});
