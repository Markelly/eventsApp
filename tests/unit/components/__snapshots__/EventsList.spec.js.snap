// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventsList/> should NOT load SortBy components 1`] = `
<div class="events-list">
  <div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`<EventsList/> should NOT load SortBy components 2`] = `
<div class="events-list">
  <div>
    <!---->
    <!---->
  </div>
</div>
`;

exports[`<EventsList/> should load empty list 1`] = `
<div class="events-list">
  <div>No events to show</div>
</div>
`;

exports[`<EventsList/> should render the component 1`] = `
<div class="events-list">
  <div>
    <sortby-stub></sortby-stub>
    <!---->
  </div>
</div>
`;
